Track Covariance Map (#814)
BLAST-ImpactX/impactx: docs/source/usage/how_to_run.rst

Run ImpactX
===========

> When to run: Physics Models

Reviewer: It's great to have this summary here, but I don't think the "when", "what", "how", headings are very helpful.  For example, when I read "When to run", my first response is:  "When I need to model something."  I suggest more direct, uniformly-phrased questions like:  "How to select a run mode", "How to select a user interface", "How to select a computing system", and "How to run the code".  I'll add some possible suggestions.

Author: Thanks! Yes, I am very flexible on these headings, thanks for the suggestions!

> When to run: Physics Models

Reviewer: ```suggestion
How to select a tracking mode
```

---------------------------

> ImpactX most powerful and central physics model is its symplectic particle tracker with collective effects (space charge, CSR, wakefields, etc.).

Reviewer: ```suggestion
ImpactX can be run using any of three distinct tracking modes.  ImpactX's most powerful tracking mode makes use of symplectic particle tracking with collective effects included (space charge, CSR, wakefields, etc.).
```

Reviewer: I would not call these "physics models".  They really are modes that differ due to what is being tracked--for example, for linear beamline elements (default) the physical model is the same in all three cases.

Author: Agreed, thx!

> Additionally, ImpactX provides two simplified physics models to aid scientists through every step, from beamline inception to operation:

Reviewer: ```suggestion
Additionally, ImpactX provides two simplified tracking modes to aid scientists through every step, from beamline inception to operation:
```

> only tracking of the reference particle orbit, or tracking of the beam envelope (covariance matrix) through linearized transport maps.

Reviewer: ```suggestion
only tracking of the reference particle orbit, or tracking of the beam envelope (6x6 covariance matrix) through linearized transport maps.
```

================== =============== =============== ==================
Mode               Use Case        Generality      Collective Effects
================== =============== =============== ==================
Particle Tracking  Full Dynamics   Most general    Supported
Reference Tracking Early Design    Reference orbit No
Envelope Tracking  Rapid Scans     Linearized      Not yet
================== =============== =============== ==================

> What to run: User Interfaces

Reviewer: ```suggestion
How to select a user interface
```

----------------------------

ImpactX can be run in three major interfaces:

#. As a Python module (script),
#. As an executable application with a key-value input file, or
#. From a :ref:`graphical user interface (dashboard) <usage-dashboard>`.

While each of these interfaces provide access to the same physics models, you might pick one or the other for specific needs, experience and workflows:

=========== ================= ========================= =========== =========== ============== ===========
Interface   Experience        Good For                  Interactive Extensible  Jupyter        HPC Support
=========== ================= ========================= =========== =========== ============== ===========
Python      Beginner-Advanced Automation, AI/ML         Yes         Yes         Yes            Yes
Application Advanced          Minimal requirements      No          No          Terminal       Yes
Dashboard   Beginner          Learning, Control-Systems Yes         No          Yes            Not yet
=========== ================= ========================= =========== =========== ============== ===========

> How to run

Reviewer: I propose moving the "How to run" section to the end of this page, and specifying that the steps described are given for a local computer.  For details on how to proceed on HPC machines, we can add a link to the existing documentation.

Author: Agreed. Let us quickly connect if you would do another change here to discuss.

Author: Resolved offline: ok for now, we will do a theory section update as a follow-up and will refine docs further.

----------

After installing ImpactX, run :ref:`one of our examples <usage-examples>`, e.g., :ref:`a FODO cell <examples-fodo>`, like this:

.. tab-set::

   .. tab-item:: Python: Script

      .. code-block:: bash

         python run_fodo.py

   .. tab-item:: Executable: Input File

      .. code-block:: bash

         impactx input_fodo.in

   .. tab-item:: Graphical User Interface

      .. code-block:: bash

         impactx-dashboard

> Where to run

Reviewer: ```suggestion
How to select a computing system
```
